wire-library/wire-protoc/src/main/java/com/squareup/wire/protocwire/Plugin.java

```diff
@@ -0,0 +1,367 @@
+    }
+
+    @Override
+    public Writer openForInsert(String filename, String insertionPoint) {
+      CodeGeneratorResponse.File.Builder file =
+        CodeGeneratorResponse.File.newBuilder();
+      file.setName(filename);
+      file.setInsertionPoint(insertionPoint);
+      return new GeneratorResponseFileWriter(file);
+    }
+
+    @Override
+    public List<FileDescriptor> getParsedFiles() {
+      return parsedFiles;
+    }
+
+    protected abstract void commitFile(CodeGeneratorResponse.File file)
+      throws IOException;
+  }
+
+  /**
+   * A {@link CodeGenerator.Context} that streams generated files to a
+   * {@link CodedOutputStream}.
+   */
+  static class StreamingContext extends AbstractContext {
+
+    private final CodedOutputStream output;
+
+    StreamingContext(List<FileDescriptor> parsedFiles,
+                            CodedOutputStream output) {
+      super(parsedFiles);
+      this.output = output;
+    }
+
+    @Override
+    protected void commitFile(File file) throws IOException {
+      // Protocol format guarantees that concatenated messages are parsed as
+      // if they had been merged in a single message prior to being serialized.
+      CodeGeneratorResponse.newBuilder().addFile(file).build().writeTo(output);
+      output.flush();
+    }
+  }
+
+  /**
+   * A {@link CodeGenerator.Context} that appends generated files to a
+   * {@link CodeGeneratorResponse}.
+   */
+  static class CodeGeneratorResponseContext extends AbstractContext {
+
+    private final CodeGeneratorResponse.Builder response;
+
+    protected CodeGeneratorResponseContext(List<FileDescriptor> parsedFiles,
+                                           CodeGeneratorResponse.Builder response) {
+      super(parsedFiles);
+      this.response = response;
+    }
+
+    @Override
+    public void commitFile(File file) {
+      response.addFile(file);
+    }
+  }
+
+  private Plugin() {
+  }
+
+  /**
+   * Runs the given code generator, reading the request from {@link System.in}
+   * and writing the response to {@link System.out}.
+   * <p>
+   * This is equivalent to {@code run(generator, new DefaultEnvironment())}.
+   *
+   * @see #run(CodeGenerator, Environment)
+   */
+  public static void run(CodeGenerator generator) throws PluginException {
+    run(generator, new DefaultEnvironment());
+  }
+
+  /**
+   * Runs the given code generator in the given environment.
+   */
+  public static void run(CodeGenerator generator, Environment environment)
+    throws PluginException {
+    CodeGeneratorRequest request;
+    try {
+      request = CodeGeneratorRequest.parseFrom(environment.getInputStream());
+    } catch (IOException e) {
+      throw new PluginException("protoc sent unparseable request to plugin.",
+        e);
+    }
+
+    List<FileDescriptor> filesToGenerate = parseFiles(request);
+    CodedOutputStream output = CodedOutputStream.newInstance(
+      environment.getOutputStream());
```

Comment:
Here, and also below around line [302](https://github.com/buildbreaker/wire/pull/1/files#diff-a68185d20cbc91b106c89d457c054ce4ed138408082fe2c0568fbb852bddec47R302), we'll want to invoke a new method on `generator` to ask it what features it supports:
```java
// new method to add to CodeGenerator interface
default List<CodeGeneratorResponse.Feature> supportedFeatures() {
    // optionally, this could return a list of ALL enum values, if we want
    // the default to be that an impl supports everything...
    return Collections.emptyList();
}
```
And then we need to write that to the output. We can go ahead and write the message to the output, even before calling anything else, because of the way the proto binary format is merge-able. (Which is also how/why the `StreamingContext` here works -- it just writes a bunch of partial messages; at the end, the union of everything written is what protoc sees.)